RetailCoder.VBE/Refactorings/Rename/RenameHandlerBase.cs

```diff
@@ -0,0 +1,106 @@
+﻿using Rubberduck.Common;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.UI;
+using Rubberduck.VBEditor;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Rubberduck.Refactorings.Rename
+{
+    public interface IRenameRefactoringHandler
+    {
+        void Rename();
+        string ErrorMessage { get; }
+    }
+
+    public class RenameHandlerBase : IRenameRefactoringHandler
+    {
+        private readonly RenameModel _model;
+        private readonly IMessageBox _messageBox;
+        private List<QualifiedModuleName> _updatedQMNs;
+
+        public RenameHandlerBase(RenameModel model, IMessageBox messageBox)
+        {
+            _model = model;
+            _messageBox = messageBox;
+            _updatedQMNs = new List<QualifiedModuleName>();
+        }
+
+        public RenameModel Model { get { return _model; } }
+
+        public IMessageBox MessageBox { get { return _messageBox; } }
+
+        public void Rewrite()
+        {
+            _updatedQMNs.Distinct().ToList()
+                .ForEach(qmn => Model.State.GetRewriter(qmn).Rewrite());
+        }
+
+        public virtual void Rename() { }
+        public virtual string ErrorMessage { get; }
+
+        public void RenameUsages(Declaration target)
+        {
+            RenameUsages(target, _model.NewName);
+        }
+
+        public void RenameUsages(Declaration target, string newName)
+        {
+            var qualifiedModuleNames = new List<QualifiedModuleName>();
+            var modules = target.References.GroupBy(r => r.QualifiedModuleName);
+            foreach (var grouping in modules)
+            {
+                qualifiedModuleNames.Add(grouping.Key);
+                var rewriter = _model.State.GetRewriter(grouping.Key);
+                var module = grouping.Key.Component.CodeModule;
+                foreach (var line in grouping.GroupBy(reference => reference.Selection.StartLine))
+                {
+                    var lastSelection = Selection.Empty;
+                    foreach (var reference in line.OrderByDescending(l => l.Selection.StartColumn))
+                    {
+                        if (reference.Selection == lastSelection)
+                        {
+                            continue;
+                        }
+                        var newContent = reference.Context.GetText().Replace(reference.IdentifierName, newName);
+                        rewriter.Replace(reference.Context, newContent);
```

Comment:
Using just newName fails for Property Setters and Getters.  When I use `newName` instead of `newContent`, in the code above, it fails because (near as I can tell...) it is replacing a full Property Get/Let statement with just the new name - it replaces too much.  When the the unit test calls GetText() to compare the results, you get a null reference exception, so you don't really get to see the result of the Rewrite() call.